### ISSUE TITLE
Node.JS V8 requires `madvise(MADV_DONTNEED)`

### ISSUE THREAD

**Issue by dimakuv** [gramineproject/gramine#1469]
### Description of the feature

Reported in https://github.com/gramineproject/gramine/discussions/1468.

Apparently Node.JS V8 has very strict error checking on `madvise()`. Here is the offending part: https://github.com/nodejs/node/blob/e7c397739ff935e8b0f54505a4e37ca711e69d94/deps/v8/src/base/platform/platform-posix.cc#L571-L575

Gramine' support for `madvise(..., MADV_DONTNEED)` is only partially implemented. And if I understand correctly, we hit this unimplemented case: https://github.com/gramineproject/gramine/blob/24319e435d6e5ec1002a5781d0f1f5362b389d32/libos/src/bookkeep/libos_vma.c#L1307

### Why Gramine should implement it?

Since it's in the common codebase of Node.JS V8, this may be an important case to add/fix in Gramine.

**Comment by jeffreyflynt** [comment]
I recently ran into this issue. With Node v10, my app loads, but I need to run my app under Node v20 and this is what I get:

```
Check failed: 0 == ret.
Fatal error in , line 0
```

Running in Docker on the latest release.

Here is the code block referenced above:

```
#else
  int ret = madvise(address, size, MADV_DONTNEED);
#endif
  // madvise with MADV_DONTNEED only fails on illegal parameters. That's a bug
  // in the caller.
  CHECK_EQ(0, ret);
  return true;
```
Any suggestions to get around this?

**Comment by dimakuv** [comment]
Just to explain why this is not so easy to implement the `MADV_DONTNEED` case for writable file-backed mmaps, here is the excerpt from the [man page of `madvise`](https://man7.org/linux/man-pages/man2/madvise.2.html):

> After a successful MADV_DONTNEED operation, the semantics of memory access in the specified region are changed: subsequent accesses of pages in the range will succeed, but will result in [...] repopulating the memory contents from the up-to-date contents of the underlying mapped file [...]

I think Gramine simply needs to invoke the `file->read()` callback, with the input buffer being the mmapped region. Unless I'm missing something, this will be trivial to implement. @kailun-qin @mkow What do you think?

For the context, this is the current Gramine code for this particular case of MADV_DONTNEED on file-backed mmaps:
https://github.com/gramineproject/gramine/blob/24319e435d6e5ec1002a5781d0f1f5362b389d32/libos/src/bookkeep/libos_vma.c#L1304-L1318

(The failing line is 1307.)

**Comment by dimakuv** [comment]
Same problem was hit here: https://github.com/gramineproject/gramine/discussions/1574

**Comment by mkow** [comment]
@dimakuv: Did you verify that it's actually that execution path? It sounds weird that it calls it on a tainted mapping, this is a rather weird feature of Linux.

**Comment by kailun-qin** [comment]
> I think Gramine simply needs to invoke the `file->read()` callback, with the input buffer being the mmapped region.

@dimakuv Did you mean to simply invoke the `file->read()` callback on `madvise(MADV_DONTNEED)` for the file-backed mappings? If so,
1. what's the expected behavior for encrypted-file-backed mappings? (after this invocation, the input buffer of the mmaped region will be updated, i.e., being overwritten if not flushed)
2. what's the correct behavior if EDMM is enabled? (do we also want to free the enclave pages)

**Comment by kailun-qin** [comment]
> I think Gramine simply needs to invoke the file->read() callback, with the input buffer being the mmapped region.

@dimakuv:
My assumption is that the issue happened w/ **private** host-backed-**non-encrypted** file mappings (because Gramine does not currently support writable pass-through file mappings on SGX, and `MAP_SHARED` is only allowed for encrypted files).

Note that there're some discussions around the behavior of `madvise(MADV_DONTNEED)` on private file mappings in Linux: https://stackoverflow.com/questions/14968309/does-madvise-madv-dontneed-instruct-the-os-to-lazily-write-to-disk.

The above proposal is to populate the memory contents from the up-to-date contents of the underlying mapped file (following the man page section cited above) at the moment of `madvise(MADV_DONTNEED)` invocation. But the man page actually mean it for **shared** file mappings and the (re-)population should happen on (later) subsequent accesses.

I'm not sure what's the right approach here and I think we should also take the behavior of when EDMM's enabled into account.

**Comment by dimakuv** [comment]
I debugged this and created a fix #1692.

I was wrong initially -- it's not about **file-backed mappings**. It's about normal **anonymous mappings**. The failing case in Gramine was this one: https://github.com/gramineproject/gramine/blob/24319e435d6e5ec1002a5781d0f1f5362b389d32/libos/src/bookkeep/libos_vma.c#L1315-L1318

Funnily, there was no reason **not to implement this support**. Apparently, we were just lazy at the time.
